Feature/reduced density matrix

Skuwar1 opened the pull request:
# New Features
- `SingleParticleExcitation` and `TwoParticleExcitation` operators are added for the FCIQMC method.

Skuwar1 committed:
21-july-2024 rDM

Skuwar1 committed:
21-july-2024 rDM

Skuwar1 committed:
21-july-2024 rDM

Skuwar1 committed:
21-july-2024 rDM2

Skuwar1 committed:
21-july-2024 rDM

Skuwar1 committed:
Update hamiltonians.md

Skuwar1 committed:
Update hamiltonians.md

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update DensityMatrixDiagonal.jl

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
````diff
@@ -0,0 +1,150 @@
+\\hat{ρ}^{(1)}_{i,j} = \\langle \\psi | \\hat a^†_{i} \\hat a_{j} | \\psi \\rangle
+```
+
+where ``i`` and ``j`` are the `mode` and ``| \\psi \\rangle`` is the `state-ket` of the given Hamiltonian.
````
The struct that you are creating does not represent the density matrix and not even its element. It represents only the operator ``\\hat a^†_{i} \\hat a_{j}``. While this is the right thing to do, I think we need another name for this, as the current name is misleading.

Maybe `SingleParticleExcitation(i, j)`?

mtsch replied:
Why are mode and state-ket in backticks? Backticks print with a monospace font and are used to refer to something in the code, i.e. function and argument names.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+        return value
+    else
+        return 0.0
+    end
```
```suggestion
    if spd.I != spd.J
        return 0.0
    end
    src = find_mode(add, spd.J)
    return src.occnum
```
For performance reasons its better to first check the equality and then avoid doing unnecessary calculations.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+struct SingleParticleReducedDensityMatrix <: AbstractHamiltonian{Float64}
+    I::Int
+    J::Int
+end
```
Maybe we should do the performance tuning in a second step, but if you are instantiating a new type instance for every matrix element, it may lead to faster code if we elevate the indices `I` and `J` to become type parameters.

mtsch replied:
I advised him against it. If you make them type parameters, it will generate a new method for each of the operators, which is very taxing on the compiler and I wouldn't expect it to bring much benefit.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+    dst = find_mode(add,spd.I)
+    if chosen == 2
+        src, dst = dst, src
+    end
```
I'm not sure why this should make sense. If you specify the operator ``a_i^\dag a_j``, why should you compute and return ``1/2(a_i^\dag a_j \pm a_j^\dag a_i)``?

Does the minus sign make any sense?

mtsch replied:
The idea here is that this shouldn't change the result, but by applying both we're increasing the probability to get a hit when computing the overlap between two sampled vectors. It would be good to check how much it actually reduces the noise though.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+    if (spd.I, spd.J) == (spd.K, spd.L)
+        return 0
+    else
+        return 6
```
Why 6?

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+end
+
+function get_offdiagonal(spd::TwoParticleReducedDensityMatrix, add::FermiFS, chosen)
+    if chosen <= 2
+        src = find_mode(add, (spd.L, spd.K))
+        dst = find_mode(add,(spd.I, spd.J))
+    elseif chosen <= 4 && chosen > 2
+        src = find_mode(add, (spd.K, spd.L))
+        dst = find_mode(add,(spd.I, spd.J))
+    else
+        src = find_mode(add, (spd.L, spd.K))
+        dst = find_mode(add,(spd.J, spd.I))
+    end
```
Why swap the pair indices? This does not change the value and brings no new information. Indeed, when you calculate overlaps with this operator more work has to be done than is needed.

joachimbrand reviewed (changes requested):
See comments below.

Skuwar1 committed:
Update ReducedDensityMatrix.jl

- Changed names of the operators
- debugged the code for off-diagonal terms in operators.

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Merge branch 'develop' into feature/ReducedDensityMatrix

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update hamiltonians.md

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update DensityMatrixDiagonal.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update ReducedDensityMatrix.jl

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+"""
+struct SingleParticleReducedDensityMatrix <: AbstractHamiltonian{Float64}
+    I::Int
+    J::Int
```
I'd prefer to use lowercase names for the indices. Uppercase is usually reserved for type names and parameters and sometimes matrices.

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+
+function get_offdiagonal(spd::SingleParticleReducedDensityMatrix, add::SingleComponentFockAddress, chosen)
+    src = find_mode(add, spd.J)
+    dst = find_mode(add,spd.I)
```
```suggestion
    dst = find_mode(add, spd.I)
```

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+
+function diagonal_element(spd::SingleParticleReducedDensityMatrix, add::SingleComponentFockAddress)
+    src = find_mode(add, spd.J)
+    dst = find_mode(add,spd.I)
```
```suggestion
    dst = find_mode(add, spd.I)
```

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+
+function diagonal_element(spd::TwoParticleReducedDensityMatrix, add::SingleComponentFockAddress)
+    src = find_mode(add, (spd.L, spd.K))
+    dst = find_mode(add,(spd.I, spd.J))
```
```suggestion
    dst = find_mode(add, (spd.I, spd.J))
```

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+function diagonal_element(spd::TwoParticleReducedDensityMatrix, add::SingleComponentFockAddress)
+    src = find_mode(add, (spd.L, spd.K))
+    dst = find_mode(add,(spd.I, spd.J))
+    address, value = excitation(add, (dst...,), (src...,))
```
This is fine, but `dst` and `src` should already be tuples:
```suggestion
    address, value = excitation(add, dst, src)
```

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+        src, dst = dst, src
+    end
+    address, value = excitation(add, (dst...,), (src...,))
+    return address, value/6
```
Please try to be consistent with formatting.
https://github.com/JuliaDiff/BlueStyle
```suggestion
    if chosen <= 2
        src = find_mode(add, (spd.L, spd.K))
        dst = find_mode(add, (spd.I, spd.J))
    elseif chosen <= 4 && chosen > 2
        src = find_mode(add, (spd.K, spd.L))
        dst = find_mode(add, (spd.I, spd.J))
    else
        src = find_mode(add, (spd.L, spd.K))
        dst = find_mode(add, (spd.J, spd.I))
    end
    if chosen % 2 == 0
        src, dst = dst, src
    end
    address, value = excitation(add, (dst...,), (src...,))
    return address, value / 6
```

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+        dst = find_mode(add,(spd.I, spd.J))
+    else
+        src = find_mode(add, (spd.L, spd.K))
+        dst = find_mode(add,(spd.J, spd.I))
```
```suggestion
        src = find_mode(add, (spd.L, spd.K))
        dst = find_mode(add,(spd.I, spd.J))
    elseif chosen <= 4 && chosen > 2
        src = find_mode(add, (spd.K, spd.L))
        dst = find_mode(add, (spd.I, spd.J))
    else
        src = find_mode(add, (spd.L, spd.K))
        dst = find_mode(add, (spd.J, spd.I))
```

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+    if chosen <= 2
+        return address, value/6
+    else
+        return address, -value/6
```
```suggestion
    if chosen % 2 == 0
        src, dst = dst, src
    end
    address, value = excitation(add, (dst...,), (src...,))
    if chosen <= 2
        return address, value / 6
    else
        return address, -value / 6
```

mtsch commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,150 @@
+"""
+    SingleParticleReducedDensityMatrix(i, j) <: AbstractHamiltonian
+
+Represent a {i,j} element of the single-particle reduced density matrix:
```
Don't like how this reads. Maybe
```suggestion
Represent the ``(i, j)``-th element of the single-particle reduced density matrix:
```

Skuwar1 replied:
Thank you, @mtsch, for your review. But it seems that you checked the old version of this pull request code before I made all the necessary changes. Please check the latest version of the code.

mtsch reviewed (changes requested):
Code looks good, but I have some comments.

joachimbrand commented on src/Hamiltonians/Hamiltonians.jl:
```diff
@@ -116,9 +119,9 @@ include("TRSymmetry.jl")
-
 include("correlation_functions.jl")
 include("DensityMatrixDiagonal.jl")
+include("ReducedDensityMatrix.jl")
```
Can we please rename this file to snake_case?
```suggestion
include("reduced_density_matrix.jl")
```
I think CamelCase should only be used for filenames if the file implements a type or module with the same name.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,104 @@
+"""
+    SingleParticleExcitation(i, j) <: AbstractHamiltonian
+
+Represent {i,j} element of the single-particle reduced density matrix:
```
```suggestion
Represent the ``{i,j}`` element of the single-particle reduced density matrix:
```

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
````diff
@@ -0,0 +1,104 @@
+\\hat{ρ}^{(1)}_{i,j} = \\hat a^†_{i} \\hat a_{j}
+```
+
+where ``i`` and ``j`` are the `mode`
````
```suggestion
where `i <: Int` and `j <: Int` specify the mode numbers.
```

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,104 @@
+    print(io, "SingleParticleExcitation($(I), $(J))")
+end
+
+LOStructure(::Type{T}) where T<:SingleParticleExcitation = AdjointUnknown()
```
```suggestion
LOStructure(::Type{<:SingleParticleExcitation}) = AdjointUnknown()
```
Just a more compact way of writing this.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,104 @@
+"""
+    TwoParticleExcitation(i, j, k, l) <: AbstractHamiltonian
+
+Represent {ij, kl} element of the two-particle reduced density matrix:
```
```suggestion
Represent the ``{ij, kl}`` element of the two-particle reduced density matrix:
```

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
````diff
@@ -0,0 +1,104 @@
+\\hat{ρ}^{(2)}_{ij, kl} =  \\hat a^†_{i} \\hat a^†_{j} \\hat a_{l} \\hat a_{k} 
+```
+
+where ``i``, ``j``, ``k``, and ``l`` are the `mode`
````
```suggestion
where `i`, `j`, `k`, and `l` (all `<: Int`) specify the mode numbers.
```

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,104 @@
+    print(io, "TwoParticleExcitation($(I), $(J), $(K), $(L))")
+end
+
+LOStructure(::Type{T}) where T<:TwoParticleExcitation = AdjointUnknown()
```
```suggestion
LOStructure(::Type{<:TwoParticleExcitation}) = AdjointUnknown()
```

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,104 @@
+    src = find_mode(add, (L, K))
+    dst = find_mode(add,(I, J))
+    address, value = excitation(add, (dst...,), (src...,))
+    if (I, J) == (K, L)
```
Is that really all? How about
```suggestion
    if (I, J) == (K, L) || (I, J) == (L, K)
```
?
Furthermore, I suggest to move the `if` condition to line 81 such that it is checked before the `find_mode` functions are executed. This will run faster for off-diagonal excitations.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,104 @@
+end
+
+function num_offdiagonals(spd::TwoParticleExcitation{I,J,K,L}, address::SingleComponentFockAddress) where {I,J,K,L}
+    if (I, J) == (K, L)
```
Ditto.

joachimbrand commented on src/Hamiltonians/ReducedDensityMatrix.jl:
```diff
@@ -0,0 +1,104 @@
+function get_offdiagonal(spd::TwoParticleExcitation{I,J,K,L}, add::SingleComponentFockAddress, chosen) where {I,J,K,L}
+    src = find_mode(add, (L, K))
+    dst = find_mode(add,(I, J))
+    address, value = excitation(add, (dst...,), (src...,))
```
See @mtsch 's comment above.

joachimbrand reviewed (changes requested):
It is much cleaner now. See comments below.

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: mtsch <matijacufar@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ReducedDensityMatrix.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update ReducedDensityMatrix.jl

Skuwar1 committed:
Update src/Hamiltonians/Hamiltonians.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Rename ReducedDensityMatrix.jl to reduced_density_matrix.jl

joachimbrand reviewed (approved):
Looks good.

mtsch reviewed (approved):
Looks good now!
